cmaes/_safe_cma.py

```diff
@@ -0,0 +1,702 @@
+            will be clipped with lower and upper bounds.
+
+        seed:
+            A seed number (optional).
+
+        population_size:
+            A population size (optional).
+
+        cov:
+            A covariance matrix (optional).
+    """
+
+    def __init__(
+        self,
+        safe_seeds: np.ndarray,
+        seeds_evals: np.ndarray,
+        seeds_safe_evals: np.ndarray,
+        safety_threshold: np.ndarray,
+        sigma: float,
+        bounds: Optional[np.ndarray] = None,
+        n_max_resampling: int = 100,
+        seed: Optional[int] = None,
+        population_size: Optional[int] = None,
+        cov: Optional[np.ndarray] = None,
+    ):
+        # safety threshold
+        self.safety_threshold = safety_threshold
+        self.safety_func_num = len(safety_threshold)
+
+        # safe seeds
+        self.safe_seeds = safe_seeds
+        self.seeds_evals = seeds_evals
+        self.seeds_safe_evals = seeds_safe_evals
+
+        n_dim = len(safe_seeds[0])
+        assert n_dim > 1, "The dimension of mean must be larger than 1"
+
+        if population_size is None:
+            population_size = 4 + math.floor(3 * math.log(n_dim))  # (eq. 48)
+        assert population_size > 0, "popsize must be non-zero positive value."
+
+        mu = population_size // 2
+
+        # hyperparameters for safe CMAES
+        self.kernel = gpytorch.kernels.RBFKernel()
+        self.kernel.lengthscale = 8.0 * n_dim
+
+        self.lip_penalty_coef = 1
+        self.lip_penalty_inc_rate = 10  # alpha
+        self.lip_penalty_dec_rate = self.lip_penalty_inc_rate ** (1.0 / n_dim)
+
+        self.lip_ite = 5  # T_data
+        self.sample_num_lip = population_size * self.lip_ite
+        self.sample_log_num = population_size * self.lip_ite
+        self.init_L_base = 10  # zeta_init
+        self.init_L = 100
+        self.gamma = 0.9
+
+        # log for safe CMAES
+        self.sampled_points = safe_seeds.copy()
+        self.sampled_safe_evals = seeds_safe_evals.copy()
+
+        # safe CMA-ES do not use negative weights
+        weights_prime = np.array(
+            np.log((population_size + 1) / 2) - np.log(np.arange(population_size) + 1)
+        )
+        weights_prime[weights_prime < 0] = 0
+
+        mu_eff = (np.sum(weights_prime[:mu]) ** 2) / np.sum(weights_prime[:mu] ** 2)
+        weights = weights_prime / weights_prime.sum()
+
+        # learning rate for the rank-one update
+        alpha_cov = 2
+        c1 = alpha_cov / ((n_dim + 1.3) ** 2 + mu_eff)
+        # learning rate for the rank-μ update
+        cmu = min(
+            1 - c1 - 1e-8,  # 1e-8 is for large popsize.
+            alpha_cov
+            * (mu_eff - 2 + 1 / mu_eff)
+            / ((n_dim + 2) ** 2 + alpha_cov * mu_eff / 2),
+        )
+        assert c1 <= 1 - cmu, "invalid learning rate for the rank-one update"
+        assert cmu <= 1 - c1, "invalid learning rate for the rank-μ update"
+
+        cm = 1  # (eq. 54)
```

Comment:
I suggest removing comments like eq. ~ (for Hansen's Tutorial paper) in this PR, as they could be confusing when compared with the equations in the SafeCMA paper.

Reply:
On the other hand, adding equation numbers for SafeCMA-specific operations would improve clarity.